erpnext/stock/stock_ledger.py

```diff
@@ -936,45 +970,100 @@ def get_sle_by_voucher_detail_no(voucher_detail_no, excluded_sle=None):
 		['item_code', 'warehouse', 'posting_date', 'posting_time', 'timestamp(posting_date, posting_time) as timestamp'],
 		as_dict=1)
 
-def get_valuation_rate(item_code, warehouse, voucher_type, voucher_no,
-	allow_zero_rate=False, currency=None, company=None, raise_error_if_no_rate=True):
+def get_valuation_rate(item_code, warehouse=None, voucher_type=None,
+	voucher_no=None, allow_zero_rate=False, currency=None, company=None,
+	raise_error_if_no_rate=True, batch_no=None):
+	"""
+		Valuation rate of an item_code is fetched, in decreasing order of
+		precedence, from the following places:
+
+			1. Last SLE having matching item_code, batch_no, warehouse;
+				field: valuation_rate
+
+			2. Last SLE having matching item_code, warehouse;
+				field: valuation_rate
+
+			3. Last SLE having matching item_code;
+				field: valuation_rate
 
+			4. Item Master;
+				field: valuation_rate
+
+			5. Item Master;
+				field: standard_rate
+
+			6. Item Price Master, buying, having matching item_code;
+				field: price_list_rate
+
+		if it isn't found in the above list then return 0.0 if allow_zero_rate.
+	"""
 	if not company:
 		company =  frappe.get_cached_value("Warehouse", warehouse, "company")
 
-	# Get valuation rate from last sle for the same item and warehouse
-	last_valuation_rate = frappe.db.sql("""select valuation_rate
-		from `tabStock Ledger Entry` force index (item_warehouse)
-		where
-			item_code = %s
-			AND warehouse = %s
-			AND valuation_rate >= 0
-			AND NOT (voucher_no = %s AND voucher_type = %s)
-		order by posting_date desc, posting_time desc, name desc limit 1""", (item_code, warehouse, voucher_no, voucher_type))
+	voucher_condition = ""
+	if voucher_no and voucher_type:
+		voucher_condition = "AND NOT (voucher_no = %(voucher_no)s AND voucher_type = %(voucher_type)s)"
+
+	query_args = dict(
+		item_code=item_code,
+		batch_no=batch_no,
+		warehouse=warehouse,
+		voucher_no=voucher_no,
+		voucher_type=voucher_type
+	)
+
+	last_valuation_rate = None
+	if batch_no and warehouse:
+		# 1. Get valuation_rate from last SLE with matching batch_no, warehouse
+		last_valuation_rate = frappe.db.sql("""SELECT valuation_rate
+			FROM `tabStock Ledger Entry` FORCE INDEX (batch_no_item_code_warehouse_index)
```

Comment:
This index isn't explicitly named, if forcing index it should be explicitly named.